src/client.rs

```diff
@@ -119,20 +151,32 @@ impl<'a> Client<'a> {
             });
 
             for _ in 0..NUM_WORKERS {
-                let idx_r = idx_r.clone();
-                let results_s = results_s.clone();
-                scope.spawn(move |_| {
-                    for idx in idx_r.iter() {
-                        match self.read_by_index(idx).map_err(Error::Err).transpose() {
-                            Some(res) => results_s.send(res).expect("results channel closed"),
-                            None => break,
+                scope.spawn({
+                    let idx_r = idx_r.clone();
+                    let settings_s = settings_s.clone();
+                    let errors_s = errors_s.clone();
+                    let stop = stop.clone();
+                    move |_| {
+                        for idx in idx_r.iter() {
+                            if stop.try_recv().is_ok() {
+                                break;
+                            }
+                            match self.read_by_index(idx) {
+                                Ok(Some(setting)) => settings_s.send(setting).unwrap(),
+                                Ok(None) => break,
+                                Err(err) => {
+                                    errors_s.send(Error::Err(err)).unwrap();
+                                    break;
+                                }
+                            }
                         }
                     }
                 });
             }
-
-            drop(results_s);
-            results_r.iter().collect::<Vec<_>>()
+            drop(idx_r);
+            drop(settings_s);
+            drop(errors_s);
```

Comment:
Why do we need these manual drops?  Doesn't the scope take care of this?

Reply:
Ah actually only need to drop the last two. For `settings_s`/`errors_s`, all the channels inside the worker threads get dropped when they exit, but the original senders are still alive. The `iter` function blocks until the channel is empty + disconnected so we have to drop the channels. I think we could also use `try_iter` which just takes everything available, but that might miss one at the end

Reply:
We can keep `idx_r` drop so we're not reliant on an implementation detail for things to work properly, a comment on why the drop is necessary would be good (basically disconnecting  our copy of the receiver from the channel so the iterator will terminate).

Reply:
Great idea, done